_posts/2016-08-22-the-java-synchronisers.md

```diff
@@ -0,0 +1,111 @@
+
+### <a name="sync-statements"></a> Synchronized statements
+
+Provides a more fine-grained synchronization mechanism. [Synchronized statements](https://docs.oracle.com/javase/tutorial/essential/concurrency/locksync.html) must specify the object that provides the intrinsic lock. 
+
+Synchronizing over separated lock objects can provide fields synchronization, without forzing synchronization between methods calls.
+
+### <a name="guarded-blocks"></a> Guarded Blocks
+
+As mentioned earlier, [guarded blocks](https://docs.oracle.com/javase/tutorial/essential/concurrency/guardmeth.html) provide support for thread coordination. Guarded blocks are part of every Java object, through `wait`, `notify` and `notifyAll` methods.
+
+The **wait** method suspend the current thread. When a thread invokes `wait`, it must own the object's intrisic lock, that is why calls to `wait` are usually wrapped in a synchronized method or statement. The invocation of the `wait` method suspends the thread execution and releases the lock. 
+
+At some point, another thread will acquire the object's intrinsic lock and invoke `notifyAll` to inform all threads waiting that something important has happened. After the second thread has released the lock, the waiting threads will reacquire the lock and resume exectuion by returning from the `wait` invocation.
+
+**Notify** wakes up a single thread. The concrete thread can not by specified, it makes the method useful only if we do not care which threads wakes up.
+
+## The Java Synchronizers
+
+Java also provide five classes for common special-purpose synchronization.
+
+### CountDownLatch
+
+The [CountDownLatch](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html) class allows one or more threads to wait until a set of operations in other threads complete. It is initialised with a count number.
+
+The `await` method **blocks** until the count reaches zero.
+The `countDown` method **decrement** the count.
+
+When the await method returns all waiting threads are released and subsequent invocations to `await` return inmidiately. The count cannot be reset.
+
+### Semaphore
+
+The [Semaphore](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html) is used to restrict threads access to a certain resource. It is initialised with a number of permits. 
+
+The `acquire` method **blocks** until a permit is available and takes it.
+The `release` method **adds** a permit, releasing a blocking acquierer. Note that calls to release does not have to be made by the same thread that called acquire.
+
+The semaphore can be **fair** or **unfair**. If fair, then the threads acquire permits in a FIFO fashion.
```

Comment:
If fair, then the threads acquire permits in a FIFO fashion. ==> When its fair, the thread acquires permits in a FIFO fashion.